dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -57,141 +53,139 @@
-  public CustomIcon getCustomIcon(String key) throws NotFoundException {
-    CustomIcon customIcon = customIconStore.getIconByKey(key);
-    if (customIcon == null) {
-      throw new NotFoundException(String.format("Icon not found: %s", key));
-    }
+  public Resource getIconResource(String key) throws NotFoundException {
 
-    return customIcon;
-  }
-
-  @Override
-  public Resource getDefaultIconResource(String key) throws NotFoundException {
-    if (defaultIcons.containsKey(key)) {
-      return new ClassPathResource(
-          String.format("%s/%s.%s", ICON_PATH, key, DefaultIcon.Icons.SUFFIX));
+    if (iconExists(key)) {
+      return new ClassPathResource(String.format("%s/%s.%s", ICON_PATH, key, DefaultIcon.SUFFIX));
     }
 
-    throw new NotFoundException(String.format("No default icon found with key %s.", key));
-  }
-
-  @Override
-  @Transactional(readOnly = true)
-  public Set<String> getKeywords() {
-    return Stream.concat(
-            defaultIcons.values().stream().map(Icon::getKeywords).flatMap(Arrays::stream),
-            customIconStore.getKeywords().stream())
-        .collect(Collectors.toSet());
+    throw new NotFoundException(String.format("No Icon found with key %s.", key));
   }
 
   @Override
   @Transactional(readOnly = true)
   public boolean iconExists(String key) {
-    return defaultIcons.get(key) != null || customIconStore.getIconByKey(key) != null;
+    return iconStore.getIconByKey(key) != null;
   }
 
   @Override
   @Transactional
-  public void addCustomIcon(CustomIcon customIcon) throws BadRequestException, NotFoundException {
-    validateIconDoesNotExists(customIcon.getKey());
-    FileResource fileResource = getFileResource(customIcon.getFileResourceUid());
-    UserDetails currentUserDetails = CurrentUserUtil.getCurrentUserDetails();
-    customIconStore.save(customIcon, fileResource, currentUserDetails);
-  }
+  public void addIcon(Icon icon) throws BadRequestException, NotFoundException, SQLException {
 
-  @Override
-  @Transactional
-  public void updateCustomIcon(String key, String description, String[] keywords)
-      throws BadRequestException, NotFoundException {
-    CustomIcon icon = validateCustomIconExists(key);
+    if (icon == null) {
+      throw new BadRequestException("Icon cannot be null.");
+    }
 
-    if (description == null && keywords == null) {
-      throw new BadRequestException(
-          String.format(
-              "Can't update icon %s if none of description and keywords are present in the request",
-              key));
+    validateIconDoesNotExists(icon);
+    validateIconKey(icon.getKey());
+
+    if (Boolean.TRUE.equals(icon.getCustom())) {
+      FileResource fileResource = getFileResource(icon.getFileResource().getUid());
+      fileResource.setAssigned(true);
+      fileResourceService.updateFileResource(fileResource);
     }
 
-    if (description != null) {
-      icon.setDescription(description);
+    if (icon.getCreatedBy() == null) {
```

Comment:
Could a user set this? Just wondering if we should always set it no matter if its null or not like with `created`.

Reply:
fixed